<a href="https://colab.research.google.com/github/sophie55555/Project/blob/main/project_1_load_digits(%EC%86%90%EA%B8%80%EC%94%A8).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

6-1. 프로젝트 (1) load_digits : 손글씨를 분류해 봅시다
우선 라이브러리 버전을 확인하고 진행할까요?

In [ ]:
import sklearn

print(sklearn.__version__)

1.5.2


(1) 필요한 모듈 import하기

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd

(2) 데이터 준비
load_digits 메서드를 사용합니다.

In [ ]:
from sklearn.datasets import load_digits

digits = load_digits() # Dictionary 자료형과 유사한 sklearn.utils.Bunch 자료형
digits.keys()

dict_keys(['data', 'target', 'frame', 'feature_names', 'target_names', 'images', 'DESCR'])

(3) 데이터 이해하기

지피지기면 백전불태! 다루어야 할 데이터를 자세히 살펴봅시다.

Feature Data 지정하기
Label Data 지정하기
Target Names 출력해 보기
데이터 Describe 해 보기

In [ ]:
digits_data = digits.data
digits_data.shape



(1797, 64)

In [ ]:
# Feature 데이터 추출
X = digits.data
print("Feature 데이터 (X):\n", X[:5]) # 처음 5개 행 출력
print("Feature 데이터 형태:", X.shape) # Feature 데이터의 형태 (행, 열) 출력

Feature 데이터 (X):
 [[ 0.  0.  5. 13.  9.  1.  0.  0.  0.  0. 13. 15. 10. 15.  5.  0.  0.  3.
  15.  2.  0. 11.  8.  0.  0.  4. 12.  0.  0.  8.  8.  0.  0.  5.  8.  0.
   0.  9.  8.  0.  0.  4. 11.  0.  1. 12.  7.  0.  0.  2. 14.  5. 10. 12.
   0.  0.  0.  0.  6. 13. 10.  0.  0.  0.]
 [ 0.  0.  0. 12. 13.  5.  0.  0.  0.  0.  0. 11. 16.  9.  0.  0.  0.  0.
   3. 15. 16.  6.  0.  0.  0.  7. 15. 16. 16.  2.  0.  0.  0.  0.  1. 16.
  16.  3.  0.  0.  0.  0.  1. 16. 16.  6.  0.  0.  0.  0.  1. 16. 16.  6.
   0.  0.  0.  0.  0. 11. 16. 10.  0.  0.]
 [ 0.  0.  0.  4. 15. 12.  0.  0.  0.  0.  3. 16. 15. 14.  0.  0.  0.  0.
   8. 13.  8. 16.  0.  0.  0.  0.  1.  6. 15. 11.  0.  0.  0.  1.  8. 13.
  15.  1.  0.  0.  0.  9. 16. 16.  5.  0.  0.  0.  0.  3. 13. 16. 16. 11.
   5.  0.  0.  0.  0.  3. 11. 16.  9.  0.]
 [ 0.  0.  7. 15. 13.  1.  0.  0.  0.  8. 13.  6. 15.  4.  0.  0.  0.  2.
   1. 13. 13.  0.  0.  0.  0.  0.  2. 15. 11.  1.  0.  0.  0.  0.  0.  1.
  12. 12.  1.  0.  0.  0.  0.  0.  1. 1

In [ ]:
# Feature 이름 추출
feature_names = digits.feature_names
print("\nFeature 이름:\n", feature_names)


Feature 이름:
 ['pixel_0_0', 'pixel_0_1', 'pixel_0_2', 'pixel_0_3', 'pixel_0_4', 'pixel_0_5', 'pixel_0_6', 'pixel_0_7', 'pixel_1_0', 'pixel_1_1', 'pixel_1_2', 'pixel_1_3', 'pixel_1_4', 'pixel_1_5', 'pixel_1_6', 'pixel_1_7', 'pixel_2_0', 'pixel_2_1', 'pixel_2_2', 'pixel_2_3', 'pixel_2_4', 'pixel_2_5', 'pixel_2_6', 'pixel_2_7', 'pixel_3_0', 'pixel_3_1', 'pixel_3_2', 'pixel_3_3', 'pixel_3_4', 'pixel_3_5', 'pixel_3_6', 'pixel_3_7', 'pixel_4_0', 'pixel_4_1', 'pixel_4_2', 'pixel_4_3', 'pixel_4_4', 'pixel_4_5', 'pixel_4_6', 'pixel_4_7', 'pixel_5_0', 'pixel_5_1', 'pixel_5_2', 'pixel_5_3', 'pixel_5_4', 'pixel_5_5', 'pixel_5_6', 'pixel_5_7', 'pixel_6_0', 'pixel_6_1', 'pixel_6_2', 'pixel_6_3', 'pixel_6_4', 'pixel_6_5', 'pixel_6_6', 'pixel_6_7', 'pixel_7_0', 'pixel_7_1', 'pixel_7_2', 'pixel_7_3', 'pixel_7_4', 'pixel_7_5', 'pixel_7_6', 'pixel_7_7']


In [ ]:
# Label 데이터 추출
y = digits.target
print("\nLabel 데이터 (y):\n", y[:5]) # 처음 5개 레이블 출력
print("Label 데이터 형태:", y.shape) # Label 데이터의 형태 출력


Label 데이터 (y):
 [0 1 2 3 4]
Label 데이터 형태: (1797,)


In [ ]:
# Label 이름 추출
target_names = digits.target_names
print("\nLabel 이름:\n", target_names)


Label 이름:
 [0 1 2 3 4 5 6 7 8 9]


(4) train, test 데이터 분리
모델 학습과 테스트용 문제지와 정답지를 준비해 봅시다.
X_train, X_test, y_train, y_test를 생성하는 방법을 참고해 보세요.

In [ ]:
# DataFrame 생성 (데이터 확인 및 분석 용이)
data = pd.DataFrame(digits.data, columns=digits.feature_names)
target = pd.DataFrame(digits.target, columns=['target'])
df = pd.concat([data, target], axis=1)

In [ ]:
# Feature와 Label 분리
X = digits.data
y = digits.target

In [ ]:
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
'''
sklearn.model_selection.train_test_split(*arrays, test_size=None, train_size=None, random_state=None, shuffle=True, stratify=None)


# test_size : 전체 데이터 셋 중 Test Dataset에 포함하려고 하는 비율
# shuffle : Training Dataset과 Test Dataset으로 나누기 전에, 데이터들을 뒤섞을지 여부
'''
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=15)

decision_tree = DecisionTreeClassifier(random_state=15)
decision_tree.fit(X_train, y_train) # 의사결정나무 모델로 학습
y_pred = decision_tree.predict(X_test) # 테스트 결과 예측

accuracy = accuracy_score(y_test, y_pred) # y_pred(답안지)를 y_test(정답지)로 채점
accuracy # 정확도 출력

0.8444444444444444

(5) 다양한 모델로 학습시켜보기

RandomForest

In [ ]:
# (1) 필요한 모듈 import
from sklearn.model_selection import train_test_split #싸이킷런에 있는 데이터를 train과 test dataset으로 나누어주는 함수 import
from sklearn.ensemble import RandomForestClassifier #랜덤포레스트라는 분류기를 사용하기 위해 import
from sklearn.metrics import classification_report #싸이킷런에 있는 분류 결과에 대한 시각화를 위해 쓰는 코드

# (2) 데이터 준비
digits = load_digits() #iris 데이터 전체를 불러온다.
digits_data = digits.data #iris데이터의 data컬럼을 분류해 iris_data 변수에 담는다.
digits_label = digits.target #iris데이터의 target컬럼을 분류해 iris_label 변수에 담는다.


X_train, X_test, y_train, y_test = train_test_split(digits_data, # iris 데이터의 data 컬럼
                                                    digits_label, # iris 데이터의 target 컬럼
                                                    test_size=0.2, # test_size : train data와 test data를 몇대몇으로 나눌지 정하는 옵션
                                                    random_state=21) # random_state : 랜덤 패턴의 값을 지정

random_forest = RandomForestClassifier(random_state=32) # RandomForest분류기 객체를 생성
random_forest.fit(X_train, y_train) # 훈련
y_pred = random_forest.predict(X_test) # 예측

print(classification_report(y_test, y_pred)) # 결과 지표를 확인

              precision    recall  f1-score   support

           0       0.97      0.94      0.95        32
           1       0.97      1.00      0.99        36
           2       1.00      1.00      1.00        30
           3       0.98      0.98      0.98        41
           4       0.94      0.97      0.95        32
           5       1.00      0.98      0.99        46
           6       1.00      0.97      0.98        32
           7       0.98      1.00      0.99        40
           8       0.93      0.98      0.95        42
           9       1.00      0.93      0.96        29

    accuracy                           0.97       360
   macro avg       0.98      0.97      0.97       360
weighted avg       0.98      0.97      0.98       360



SVM 모델은 다음과 같이 사용합니다.

In [ ]:
from sklearn import svm #Support Vector Machine을 사용하기 위해 import
svm_model = svm.SVC() # 모델 객체를 만든다.

print(svm_model._estimator_type) # 이 모델이 어떤 타입인지 확인한다.

svm_model.fit(X_train, y_train) # 훈련
y_pred = svm_model.predict(X_test) # 예측

print(classification_report(y_test, y_pred)) # 결과 지표를 확인

classifier
              precision    recall  f1-score   support

           0       1.00      0.97      0.98        32
           1       0.95      1.00      0.97        36
           2       1.00      1.00      1.00        30
           3       1.00      1.00      1.00        41
           4       0.97      1.00      0.98        32
           5       0.98      1.00      0.99        46
           6       1.00      1.00      1.00        32
           7       1.00      1.00      1.00        40
           8       0.98      0.95      0.96        42
           9       1.00      0.93      0.96        29

    accuracy                           0.99       360
   macro avg       0.99      0.99      0.99       360
weighted avg       0.99      0.99      0.99       360



SGD Classifier 모델은 다음과 같이 사용합니다.

In [ ]:
from sklearn.linear_model import SGDClassifier #선형분류기인 SGDClassifier를 사용하기 위한 import
sgd_model = SGDClassifier() # 모델 객체 생성

print(sgd_model._estimator_type) # 이 모델의 타입을 확인

sgd_model.fit(X_train, y_train) # sgd모델로 훈련데이터로 훈련시킨다.
y_pred = sgd_model.predict(X_test)# 그 모델로 test데이터를 사용해 예측

print(classification_report(y_test, y_pred)) # 결과 지표를 확인

classifier
              precision    recall  f1-score   support

           0       1.00      0.97      0.98        32
           1       0.97      0.86      0.91        36
           2       1.00      1.00      1.00        30
           3       0.95      0.98      0.96        41
           4       0.97      1.00      0.98        32
           5       0.98      0.98      0.98        46
           6       1.00      1.00      1.00        32
           7       1.00      1.00      1.00        40
           8       0.85      0.98      0.91        42
           9       1.00      0.90      0.95        29

    accuracy                           0.97       360
   macro avg       0.97      0.97      0.97       360
weighted avg       0.97      0.97      0.97       360



Logistic Regression 모델은 다음과 같이 사용합니다.

In [ ]:
from sklearn.linear_model import LogisticRegression # 선형분류기인 LogisticRegression를 사용하기 위한 import
logistic_model = LogisticRegression() # 모델 객체 생성

print(logistic_model._estimator_type) # 이 모델의 타입을 확인

logistic_model.fit(X_train, y_train) #LogisticRegression모델로 훈련데이터를 가지고 훈련시킨다.
y_pred = logistic_model.predict(X_test) # 예측

print(classification_report(y_test, y_pred)) # 결과 지표를 확인

classifier
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        32
           1       0.95      0.97      0.96        36
           2       1.00      1.00      1.00        30
           3       0.98      1.00      0.99        41
           4       0.94      0.97      0.95        32
           5       0.98      0.98      0.98        46
           6       1.00      1.00      1.00        32
           7       0.97      0.97      0.97        40
           8       1.00      0.95      0.98        42
           9       0.96      0.93      0.95        29

    accuracy                           0.98       360
   macro avg       0.98      0.98      0.98       360
weighted avg       0.98      0.98      0.98       360



/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


(6) 모델을 평가해 보기
학습된 모델들의 테스트데이터 예측 결과를 어떻게 해석해야 할까요? 모델의 성능을 평가하는 지표로는 무엇이 좋을까요? sklearn.metrics 에서 제공하는 평가지표 중 적절한 것을 선택해 보세요. 선택하신 이유도 설명해 주세요.

In [ ]:
fake_pred = [0] * len(y_pred) # 0으로만 이루어진 가짜 답안지 준비

accuracy = accuracy_score(y_test, fake_pred) # fake_pred(가짜 답안지)를 y_test(정답지)로 채점
accuracy # 정확도 출력

0.08888888888888889

In [ ]:
from sklearn.metrics import confusion_matrix

confusion_matrix(y_test, y_pred)

array([[32,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0, 35,  0,  1,  0,  0,  0,  0,  0,  0],
       [ 0,  0, 30,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0, 41,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0, 31,  0,  0,  1,  0,  0],
       [ 0,  0,  0,  0,  1, 45,  0,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0, 32,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0, 39,  0,  1],
       [ 0,  1,  0,  0,  1,  0,  0,  0, 40,  0],
       [ 0,  1,  0,  0,  0,  1,  0,  0,  0, 27]])

In [ ]:
confusion_matrix(y_test, fake_pred)

array([[32,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [36,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [30,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [41,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [32,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [46,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [32,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [40,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [42,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [29,  0,  0,  0,  0,  0,  0,  0,  0,  0]])

In [ ]:
from sklearn.metrics import classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        32
           1       0.95      0.97      0.96        36
           2       1.00      1.00      1.00        30
           3       0.98      1.00      0.99        41
           4       0.94      0.97      0.95        32
           5       0.98      0.98      0.98        46
           6       1.00      1.00      1.00        32
           7       0.97      0.97      0.97        40
           8       1.00      0.95      0.98        42
           9       0.96      0.93      0.95        29

    accuracy                           0.98       360
   macro avg       0.98      0.98      0.98       360
weighted avg       0.98      0.98      0.98       360



In [ ]:
print(classification_report(y_test, fake_pred, zero_division=0))

              precision    recall  f1-score   support

           0       0.09      1.00      0.16        32
           1       0.00      0.00      0.00        36
           2       0.00      0.00      0.00        30
           3       0.00      0.00      0.00        41
           4       0.00      0.00      0.00        32
           5       0.00      0.00      0.00        46
           6       0.00      0.00      0.00        32
           7       0.00      0.00      0.00        40
           8       0.00      0.00      0.00        42
           9       0.00      0.00      0.00        29

    accuracy                           0.09       360
   macro avg       0.01      0.10      0.02       360
weighted avg       0.01      0.09      0.01       360



In [ ]:
accuracy_score(y_test, y_pred), accuracy_score(y_test, fake_pred)

(0.9777777777777777, 0.08888888888888889)

(모델 평가)

RandomForest -  
정확도가 97%이고, 대부분의 숫자에서 정밀도, 재현율, F1 점수가 90% 이상으로 매우 높습니다. 이는 RandomForest 모델이 손으로 쓴 숫자를 매우 잘 구별한다는 것을 의미합니다.

svm - 99%의 정확도

SGD Classifier - 97%의 정확도

Logistic Regression - 98%


모든 모델들이 대체적으로 높은 정확도를 보여주었음.

accuracy_score(y_test, y_pred): 모델의 정확도
accuracy_score(y_test, fake_pred): 가짜 예측의 정확도

모델의 정확도가 97.8%로 매우 높다.
가짜 예측 즉,모든 예측을 특정 값 (예: 0)으로 했을 때의 정확도가 약 8.9%로 매우 낮음.

현재 모델의 성능이 우수하다고 판단할 수 있음.